code/service/src/main/scala/com/namely/chiefofstate/ScalapbSerializer.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2020 Namely Inc.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+package com.namely.chiefofstate
+
+import akka.actor.ExtendedActorSystem
+import akka.serialization.BaseSerializer
+import scalapb.{GeneratedMessage, GeneratedMessageCompanion}
+
+import java.util.concurrent.atomic.AtomicReference
+
+sealed class ScalapbSerializer(val system: ExtendedActorSystem) extends BaseSerializer {
+  private val classToCompanionMapRef = new AtomicReference[Map[Class[_], GeneratedMessageCompanion[_]]](Map.empty)
+  override def toBinary(o: AnyRef): Array[Byte] = {
+    o match {
+      case e: GeneratedMessage => e.toByteArray
+      case _                   => throw new IllegalArgumentException("need a subclass of scalapb.GeneratedMessage")
+    }
+  }
+
+  override def includeManifest: Boolean = true
+
+  override def fromBinary(bytes: Array[Byte], manifest: Option[Class[_]]): AnyRef = {
+    manifest match {
+      case Some(clazz) =>
+        @scala.annotation.tailrec
+        def messageCompanion(companion: GeneratedMessageCompanion[_] = null): GeneratedMessageCompanion[_] = {
+          val classToCompanion = classToCompanionMapRef.get()
+          classToCompanion.get(clazz) match {
+            case Some(cachedCompanion) => cachedCompanion
+            case None =>
+              val uncachedCompanion =
+                if (companion eq null)
+                  Class
+                    .forName(clazz.getName + "$", true, clazz.getClassLoader)
```

Comment:
I would prefer we avoid this sort of dynamic class reference. let's just user protobuf as it's meant to be used and map type URL's to companions.

Reply:
This was to show you what Nadav came up with.